trellis/containers/transforms.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2022 Agtonomy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "transforms.hpp"
+
+#include <sstream>
+
+namespace trellis {
+namespace containers {
+
+void Transforms::UpdateTransform(const std::string& from, const std::string& to, const RigidTransform& transform,
+                                 std::chrono::milliseconds validity_window) {
+  UpdateTransform(from, to, transform, validity_window, core::time::Now());
+}
+
+void Transforms::UpdateTransform(const std::string& from, const std::string& to, const RigidTransform& transform,
+                                 std::chrono::milliseconds validity_window,
+                                 const trellis::core::time::TimePoint& when) {
+  Insert(from, to, transform, validity_window, when);
+  // We will derive and insert the inverse transform as well
+  Insert(to, from, transform.Inverse(), validity_window, when);
+}
+
+bool Transforms::HasTransform(const std::string& from, const std::string& to) const {
+  return HasTransform(from, to, core::time::Now());
+}
+
+bool Transforms::HasTransform(const std::string& from, const std::string& to,
+                              const trellis::core::time::TimePoint& when) const {
+  const auto timestamp = FindNearestTransformTimestamp(from, to, when);
+  return !timestamp ? false : true;
+}
+
+const Transforms::RigidTransform& Transforms::GetTransform(const std::string& from, const std::string& to) const {
+  return GetTransform(from, to, core::time::Now());
+}
+
+const Transforms::RigidTransform& Transforms::GetTransform(const std::string& from, const std::string& to,
+                                                           const trellis::core::time::TimePoint& when) const {
+  const auto timestamp = FindNearestTransformTimestamp(from, to, when);
+
+  if (!timestamp) {
+    std::stringstream msg;
+    msg << "No transform found for " << from << " -> " << to << " at " << trellis::core::time::TimePointToSeconds(when);
+    throw std::runtime_error(msg.str());
+  }
+
+  // Our value is guaranteed to exist at this point
+  const auto& transform_map = transforms_.at(CalculateKeyFromFrames(from, to));
+  return transform_map.at(*timestamp).transform;
```

Comment:
What happens if the transform requested is chained. For example, if the map has A->B, and B->C, and the GetTransform request is for A->C, we should be able to compute the resultant Transform from operations on the two in the map.